core/src/main/java/org/hibernate/ogm/dialect/batch/spi/OperationsQueue.java

```diff
@@ -40,14 +43,34 @@ public boolean isClosed() {
 
 	private final Queue<Operation> operations = new LinkedList<Operation>();
 
-	private final Set<EntityKey> entityKeys = new HashSet<EntityKey>();
+	private final Map<EntityKey, GroupedChangesToEntityOperation> groupedOperations = new HashMap<>();
+
+	private final Set<EntityKey> insertionQueue = new HashSet<>();
 
 	private boolean closed = false;
 
+	public OperationsQueue() {
+	}
+
 	public void add(InsertOrUpdateTupleOperation operation) {
+		add( (GroupableEntityOperation) operation );
+		insertionQueue.add( operation.getEntityKey() );
+	}
+
+	public void add(GroupableEntityOperation operation) {
```

Comment:
Yes, you're right. I did that because I'm not a big fan of the instanceof constructs but it's definitely not a good idea. That being said, considering we must have the dispatching in `add(Operation)`, I removed the other methods: I think it's clearer like that.